hoomd/md/tune/nlist_buffer.py

```diff
@@ -0,0 +1,322 @@
+            self._last_tps = tps
+            if tps > self._max_tps:
+                self._best_buffer_size = self._tunable.x
+                self._max_tps = tps
+        if not self.tuned:
+            if self.solver.solve([self._tunable]):
+                self._tuned += 1
+            else:
+                self._tuned = 0
+
+    def _buffer_post(self, value: float):
+        self._update_buffer_domain = True
+        return value
+
+    def attach(self, simulation):
+        self._simulation = simulation
+        self._tunable = self._make_tunable(self.nlist)
+
+    def _make_tunable(self, nlist):
+        return hoomd.tune.ManualTuneDefinition(
+            get_y=_IntervalTPS(self._simulation),
+            target=0.0,
+            get_x=self._get_buffer,
+            set_x=self._set_buffer,
+            domain=(0, self.maximum_buffer),
+        )
+
+    def _get_buffer(self):
+        return self.nlist.buffer
+
+    def _set_buffer(self, new_buffer):
+        self.nlist.buffer = new_buffer
+
+    def detach(self):
+        self._simulation = None
+        self._tunable = None
+
+    @property
+    def tuned(self):
+        """bool: Whether the neighbor list buffer is considered tuned.
+
+        The tuner is considered tune when the specified solver returns ``True``
+        when solving twice consecutively. See `hoomd.tune` for more information
+        on tuning criteria.
+        """
+        return self._tuned > 1
+
+    @hoomd.logging.log
+    def max_tps(self):
+        """int: The maximum recorded TPS during tuning."""
+        return self._max_tps
+
+    @hoomd.logging.log
+    def best_buffer_size(self):
+        """float: The buffer size corresponding to ``max_tps`` during tuning."""
+        return self._best_buffer_size
+
+    @hoomd.logging.log
+    def last_tps(self):
+        """int: The last TPS computed for the tuner."""
+        return self._last_tps
+
+    @property
+    def final_bin(self):
+        """tuple[float, float]: Boundaries of grid search optimization.
+
+        Property is only available when a `hoomd.tune.GridOptimizer` is used,
+        and tuning is finished. Otherwise, the property is ``None``.
+        """
+        if not isinstance(self.solver, hoomd.tune.GridOptimizer):
+            return None
+        if self._tunable is None:
+            return None
+        return self.solver._final_bins.get(self._tunable, None)
+
+    def __getstate__(self):
+        state = copy.copy(self.__dict__)
+        for attr in self._skip_for_equality:
+            state.pop(attr, None)
+        return state
+
+
+class NeighborListBuffer(hoomd.tune.custom_tuner._InternalCustomTuner):
+    """Optimize neighbor list buffer size for maximum TPS.
+
+    Direct instantiation of this class requires a `hoomd.tune.Optimizer`
+    that determines how move sizes are updated. This class also provides class
+    methods to create a `NeighborListBuffer` tuner with built-in solvers; see
+    `NeighborListBuffer.with_grid` and
```

Comment:
Could we put this in a `Tip:` block to make it more obvious? I've gotten feedback from some users that tried (and failed) to use the MC move size tuners with the documented constructor and missed the helpful `with_` classmethods.